### PR TITLE
WIP - Netdata v2

### PR DESCRIPTION
**Pull request by ktsaou**
1. **Dashboards v0 and v1 are removed**
   These dashboards are not maintained for a long time and probably they are a today a liability for the security of Netdata.

   - [ ] Remove dashboard v0
   - [ ] Remove dashboard v1

2. **New Netdata Cloud claimimng process**

   - [x] Implement claiming internally in Netdata (via `/etc/netdata/claim.conf`)
   - [x] To support this, require `libcurl` in Netdata
   - [x] Remove `netdata-claim.sh`
   - [x] Customize the messages (`cmd`, `help`) in the API responsing or `/api/v1/claim` to support Windows

3. **Netdata Cloud SSO should work for air-gapped children**

   - [x] Implement NODEID streaming capability, for parents to send back to children their node id and netdata cloud base URL
   - [x] Convert bearer token API into a function (allowing indirectly claimed agents to generate bearer tokens)
   - [x] Use the new bearer token function to acquire bearer token via Netdata Cloud

4. **Code re-organization and cleanup**

   - [x] Remove all options for building an agent without Netdata Cloud or disabling Netdata Cloud
      - [x] remove `ENABLE_CLOUD` - netdata can only be built with cloud enabled
      - [x] remove `ENABLE_ACLK` - netdata can only be built with aclk enabled
      - [x] remove `DISABLE_CLOUD` - netdata can only be built with cloud
      - [x] remove `NETDATA_DISABLE_CLOUD`.
      - [x] remove `ENABLE_HTTPS`
      - [x] remove `ENABLE_OPENSSL`
      - [x] remove `netdata_cloud_enabled`
   - [x] Remove all options for disabling Netdata Cloud from the installers:
      - [x] remove `--disable-cloud`
      - [x] remove `--require-cloud`
      - [x] remove `--disable-https`
   - [x] Split the APIs by version to make the code more manageable

5. **Windows Support**

   - [x] Support external plugins in the Windows version of Netdata (new spawn server compatible with Windows)

|Server|Description|Linux|FreeBSD|MacOS|CygWin<br/>Msys2|Windows
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
NoFork|Forks at the beginning<br/>then transfers pipe fds to new process|Yes|Yes|Yes|No<br/>No support for transfering fds|No|
Posix|Uses `posix_spawn()`<br/>to create new processes|Yes|?|?|Yes|No|
libuv|Uses an even loop per spawn server<br/> and uses `uv_spawn()` to create new processes|Yes|?|?|No|Probably|
Windows|Uses `CreateProcess()` to spawn new processes|No|No|No|Yes|Probably|

6. **New v3 dashboard**

   - [ ] `/index.html` is a new splash screen that automatically takes care of users' signin-in, agents' claiming and loading dashboard v3
   - [ ] `/index.html` provides the option to use the old v2 dashboard, with a warning that this will be removed in the future
   - [ ] new v3 dashboard that can exclusively be accessed from logged-in users on (directly or indirectly) claimed agents
   - [ ] new v3 dashboard uses the new TOC for metrics by default
   - [ ] new v3 dashboard on the agent, uses the real Netdata Cloud space for user settings and themes
   - [ ] new v3 dashboard inherits the dashboard limits Netdata Cloud has
   - [x] new v3 dashboard uses the new claiming fields to properly instruct Windows users on finding the private key required via claiming from the UI

7. **Old v2 dashboard, will have the 5-nodes limit Netdata Cloud Community has**
   With this release, the dashboard v2 become obsolete and will no longer be maintained in the future. It will still be available though for some time.

   - [ ] The existing v2 dashboard will now have the limit of 5 nodes per dashboard, on non-signed in users and community plan users
   - [x] new API `/api/v3/settings` supports saving dashboard settings at Netdata. 

8. **License Manager**
   
   - [x] In `netdata.conf` at the `[cloud]` section, add the option `mode = normal | license manager`.
      The `license manager` mode, has the following effects:

      - [x] No contexts or alerts are sent to Netdata Cloud (streaming and snapshots)
      - [x] No data queries or functions are allowed from Netdata Cloud, except bearer token acquiring

   When claiming is done via a parent at a DMZ, the user can specify `send charts matching = !*` in `stream.conf`of the children, to prevent streaming metrics to the parent.

9. **Protected API**

	- [x] Add a netdata.conf option to protect the API with a bearer token
	- [x] Bearer tokens should be saved and loaded back
	- [x] `/api/v3/me` returns information about the current user
	- [ ] front-end: reloading the dashboard should reuse the last bearer token, if available.

10. **Other Changes**

   - [x] systemd-journal plugin now respects user supplied facets (required for supporting themes in Netdata UI v3)
   - [x] systemd-journal plugin now accepts POST payload

---

## Tests needed

- [x] claim an agent via kickstart.sh (dynamic install)
- [x] claim an agent via kickstart.sh (static install)
- [x] claim an agent via kickstart.sh (source install)
- [x] claim an agent via Netdata UI
- [x] claim an agent via `/etc/netdata/claim.conf`
- [x] Unclaimed child becomes indirectly claimed if a parent is claimed (access logs to verify it works)
- [x] Unclaimed child becomes indirectly claimed if a grand-parent is claimed (access logs to verify it works)

### CODE REVIEW

**Comment by Ferroin** [comment]
Note that failure for the ‘Test Consolidated Artifacts (Source)’ check is expected due to the fact that it uses the production version of `install-required-packages.sh` instead of the copy from the PR.

**Comment by papazach** [comment]
@ktsaou [here](https://github.com/netdata/cloud-agent-data-ctrl-service/blob/master/transport/http.go#L47) you can find a list of the Agent endpoints used (legacy ones included as they're still supported)

**Comment by sashwathn** [comment]
@ktsaou : Is this PR handling the restriction of `5 nodes` on their parent dashboard.

File: src/claim/netdata-claim.sh.in [hunk 1]
```diff
@@ -1,451 +1,109 @@
...
-        fi
-    fi
-    if [ "${NETDATA_RUNNING}" -eq 0 ]; then
-        config_result=$(@sbindir_POST@/netdata 2>/dev/null -W get2 "$conf_file" "$section" "$key_name" unknown_default)
+    if [ -n "${override}" ]; then
+        echo "${override}"
+    elif (echo "${value}" | grep -q '@'); then
+        echo "${default}"
+    else
+        echo "${value}"
```
Comment (ilyam8) [inline src/claim/netdata-claim.sh.in:20]: This returns an empty string when:

- no override
- `$1` is empty

Comment (ilyam8) [reply src/claim/netdata-claim.sh.in:20]: I did install from source and for some reason `@netdata_group_POST@` is an empty string.

Comment (Ferroin) [reply src/claim/netdata-claim.sh.in:20]: This should be fixed now, the original handling of the templating assumed changes that I forgot were not yet in the master branch.

File: packaging/installer/kickstart.sh [hunk 2]
```diff
@@ -1277,6 +1235,93 @@ is_netdata_running() {
...
+  fi
+  if [ -n "${NETDATA_CLAIM_INSECURE}" ]; then
+      config="${config}\n    insecure = ${NETDATA_CLAIM_INSECURE}"
+  fi
+
+  run_as_root touch "${claim_config}.tmp" || return 1
+  run_as_root chmod 0640 "${claim_config}.tmp" || return 1
+  run_as_root chown ":${NETDATA_CLAIM_GROUP}" "${claim_config}.tmp" || return 1
+  run_as_root echo "${config}" > "${claim_config}.tmp" || return 1
+  run_as_root mv -f "${claim_config}.tmp" "${claim_config}" || return 1
```
Comment (ilyam8) [inline packaging/installer/kickstart.sh:1279]: - `"${NETDATA_CLAIM_GROUP}` is not set
- `claim.conf` will be root:root 640 (not readable by `netdata`)

File: packaging/installer/kickstart.sh [hunk 3]
```diff
@@ -1326,34 +1372,16 @@ claim() {
...
-    2) warning "Unable to claim node due to issues creating the claiming directory or preparing the local claiming key. Make sure you have a working openssl command and that ${INSTALL_PREFIX}/var/lib/netdata/cloud.d exists, then try again." ;;
-    3) warning "Unable to claim node due to missing dependencies. Usually this means that the Netdata Agent was built without support for Netdata Cloud. If you built the agent from source, please install all needed dependencies for Cloud support. If you used the regular installation script and see this error, please file a bug report at ${AGENT_BUG_REPORT_URL}." ;;
-    4) warning "Failed to claim node due to inability to connect to ${NETDATA_CLAIM_URL}. Usually this either means that the specified claiming URL is wrong, or that you are having networking problems." ;;
-    5)
-      progress "Successfully claimed node, but was not able to notify the Netdata Agent. You will need to restart the Netdata service on this node before it will show up in the Cloud."
-      return 0
+  case ${method} in
+    script) run_claim_script ;;
+    internal)
```
[the remark this answers is not in the document]
Comment (ilyam8) [reply packaging/installer/kickstart.sh:1380]: And I think we need `*` case here

File: packaging/installer/kickstart.sh (hunk 2 above)
Comment (Ferroin) [reply packaging/installer/kickstart.sh:1279]:
```suggestion
  run_as_root touch "${claim_config}.tmp" || return 1
  run_as_root chmod 0640 "${claim_config}.tmp" || return 1
  run_as_root chown ":${NETDATA_CLAIM_GROUP:-netdata}" "${claim_config}.tmp" || return 1
  run_as_root echo "${config}" > "${claim_config}.tmp" || return 1
  run_as_root mv -f "${claim_config}.tmp" "${claim_config}" || return 1
```

**Pull request merged**